syntax = "proto3";

package mpointers;

service MemoryManager {
  rpc Create(CreateRequest) returns (CreateResponse) {}
  rpc Set(SetRequest) returns (SetResponse) {}
  rpc Get(GetRequest) returns (GetResponse) {}
  rpc IncreaseRefCount(RefCountRequest) returns (RefCountResponse) {}
  rpc DecreaseRefCount(RefCountRequest) returns (RefCountResponse) {}
}

message CreateRequest {
  int32 size = 1;
  string type = 2;
}

message CreateResponse {
  int32 id = 1;
  bool success = 2;
  string error_message = 3;
}

message SetRequest {
  int32 id = 1;
  bytes value = 2;
}

message SetResponse {
  bool success = 1;
  string error_message = 2;
}

message GetRequest {
  int32 id = 1;
}

message GetResponse {
  bytes value = 1;
  bool success = 2;
  string error_message = 3;
}

message RefCountRequest {
  int32 id = 1;
}

message RefCountResponse {
  bool success = 1;
  string error_message = 2;
}